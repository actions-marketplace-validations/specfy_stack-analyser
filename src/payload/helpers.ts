import { listIndexed } from '../common/techs';
import type { AllowedKeys } from '../types/techs';

import { Payload } from '.';

/**
 * When receive a tech in a component, we can deduct a new Component that was missing
 * For example we receive:
 *  - "pg" from a package.json
 *  - we deduct there is a Postgresql
 *  - Create an eponymous component
 *
 * Obviously there could be some false positive.
 */
export function findImplicitComponent(pl: Payload, tech: AllowedKeys) {
  const ref = listIndexed[tech];
  if (ref.type === 'ci' || ref.type === 'language' || ref.type === 'tool') {
    return;
  }

  const comp = new Payload({
    name: tech,
    tech: tech,
    folderPath: pl.path[0],
    parent: pl,
  });
  pl.addComponent(comp);
  if (comp.group !== 'hosting') {
    pl.addEdges(comp.id);
  }
}

/**
 * When receive a tech in a component, we can deduct a new Hosting that was missing
 * For example we receive:
 *  - "vercel" from a folder named .vercel
 *  - we deduct this component is hosted by Vercel
 *  - Create an eponymous component
 *
 * Obviously there could be some false positive.
 */
export function findHosting(pl: Payload, tech: AllowedKeys) {
  const ref = listIndexed[tech];
  if (ref.type !== 'hosting') {
    return;
  }

  const find = pl.components.find((c) => c.tech === ref.key);
  if (!find) {
    throw new Error(`cant find hosting ${ref.key}`);
  }

  pl.inComponent = find.id;
}

export function findEdges(pl: Payload) {
  pl.components.forEach((component) => {
    component.techs.forEach((tech) => {
      const ref = listIndexed[tech];

      if (
        ref.type === 'sass' ||
        ref.type === 'app' ||
        ref.type === 'db' ||
        ref.type === 'messaging'
      ) {
        const find = pl.components.find((c) => c.tech === ref.key);
        if (!find) {
          throw new Error(`cant find sass ${ref.key}`);
        }
        component.addEdges(find.id);

        return;
      }
    });
  });
}
