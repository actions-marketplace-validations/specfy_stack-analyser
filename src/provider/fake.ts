import path from 'node:path';

import type { BaseProvider, ProviderFile } from './base';

export class FakeProvider implements BaseProvider {
  private paths;
  private files;

  constructor(opts: {
    paths: Record<string, string[]>;
    files: Record<string, string>;
  }) {
    this.paths = opts.paths;
    this.files = opts.files;
  }

  listDir(pathRelative: string): Promise<ProviderFile[]> {
    if (!(pathRelative in this.paths)) {
      return Promise.resolve([]);
    }

    const files = this.paths[pathRelative].sort();
    return Promise.resolve(
      files.map((file) => {
        return {
          name: file,
          type: file.endsWith('/') ? 'dir' : 'file',
          fp: path.join(pathRelative, file),
        };
      })
    );
  }

  open(pathRelative: string): Promise<string> {
    if (!(pathRelative in this.files)) {
      throw new Error(`not found "${pathRelative}"`);
    }

    return Promise.resolve(this.files[pathRelative]);
  }
}
