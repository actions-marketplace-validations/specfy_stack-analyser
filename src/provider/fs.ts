import fs from 'fs/promises';
import path from 'node:path';

import type { BaseProvider, ProviderFile } from './base';

export interface FSProviderOptions {
  path: string;
  ignorePaths?: Array<RegExp | string>;
}

export class FSProvider implements BaseProvider {
  opts;

  constructor(opts: FSProviderOptions) {
    this.opts = opts;
  }

  async listDir(pathRelative: string): Promise<ProviderFile[]> {
    const fp = path.join(this.opts.path, pathRelative);

    const files = await fs.readdir(fp, {
      withFileTypes: true,
    });

    const list: ProviderFile[] = files.map((file) => {
      return {
        name: file.name,
        type: file.isDirectory() ? 'dir' : 'file',
        fp: path.join(pathRelative, file.name),
      };
    });

    return list;
  }

  async open(pathRelative: string): Promise<string> {
    const fp = path.join(this.opts.path, pathRelative);
    const content = await fs.readFile(fp);

    return content.toString();
  }
}
