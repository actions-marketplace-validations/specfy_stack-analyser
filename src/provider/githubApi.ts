import path from 'node:path';

import type { Octokit } from 'octokit';

import type { BaseProvider, ProviderFile } from './base.js';

export interface GithubAPIProviderOptions {
  path: string;
  octokit: Octokit;
  owner: string;
  repo: string;
  hash: string;
  ignorePaths?: Array<RegExp | string>;
}

export class GithubAPIProvider implements BaseProvider {
  opts;
  basePath;
  private tree: ProviderFile[] | null = null;

  constructor(opts: GithubAPIProviderOptions) {
    this.opts = opts;
    this.basePath = opts.path;
  }

  async listDir(pathRelative: string): Promise<ProviderFile[]> {
    if (!this.tree) {
      await this.getTree();
    }

    return this.tree!.filter((item) => {
      return path.dirname(item.fp) === pathRelative;
    });
  }

  async open(pathRelative: string): Promise<string> {
    const res = await this.opts.octokit.rest.repos.getContent({
      owner: this.opts.owner,
      repo: this.opts.repo,
      tree_sha: this.opts.hash,
      path: pathRelative,
    });

    if (!('content' in res.data)) {
      throw new Error('no content in data');
    }

    const decoded = Buffer.from(res.data.content, 'base64').toString();
    return decoded;
  }

  private async getTree() {
    const res = await this.opts.octokit.rest.git.getTree({
      owner: this.opts.owner,
      repo: this.opts.repo,
      tree_sha: this.opts.hash,
      recursive: 'true',
    });

    const tree: ProviderFile[] = res.data.tree.map((item) => {
      if (!item.path) {
        throw new Error('no path');
      }

      const name = path.basename(item.path);

      return {
        name,
        type: item.type === 'blob' ? 'file' : 'dir',
        fp: path.join('/', item.path),
      };
    });

    this.tree = tree;
  }
}
