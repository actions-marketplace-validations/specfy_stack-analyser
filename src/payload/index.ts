import path from 'node:path';

import { languages, others } from '../common/languages';
import { nid } from '../common/nid';
import { nameToKey } from '../common/techs';
import type { BaseProvider } from '../provider/base';
import { IGNORED_DIVE_PATHS } from '../provider/base';
import { rulesComponents, rulesTechs } from '../rules';
import type { ComponentGroup, GraphEdge, TechAnalyser } from '../types';
import type { AllowedKeys } from '../types/techs';

import '../rules/index';

export class Payload {
  public id: string;
  public languages: Record<string, number>;
  public components: Payload[];
  public path: string;
  public name: string;
  public group: ComponentGroup;
  public techs: Set<AllowedKeys>;
  public inComponent: string | null;
  public tech: AllowedKeys | null;

  private edges: GraphEdge[];

  constructor({
    name,
    folderPath,
    tech,
    group,
  }: {
    name: string;
    folderPath: string;
    tech?: AllowedKeys | null;
    group?: ComponentGroup;
  }) {
    this.id = nid();
    this.name = name;
    this.path = folderPath;
    this.tech = tech || null;
    this.edges = [];
    this.group = group || 'component';
    this.inComponent = null;

    this.components = [];
    this.techs = new Set();
    this.languages = {};
  }

  async recurse(provider: BaseProvider, filePath: string) {
    const files = await provider.listDir(filePath);

    let ctx: Payload = this;
    for (const rule of rulesComponents) {
      const res = await rule(files, provider);
      if (!res) {
        continue;
      }

      if (res.name !== 'virtual') {
        ctx = res;
      } else {
        ctx.merge(res);
      }
    }

    // Detect Tech
    for (const rule of rulesTechs) {
      const res = rule(files);
      if (!res) {
        continue;
      }

      ctx.addTech(res.key);
    }

    // Recursively dive in folders
    for (const file of files) {
      if (file.type === 'file') {
        ctx.detectLang(file.name);
        continue;
      }
      if (IGNORED_DIVE_PATHS.includes(file.name)) {
        continue;
      }

      const fp = path.join(filePath, file.name);

      await ctx.recurse(provider, fp);
    }

    if (ctx !== this) {
      this.merge(ctx);
    }
  }

  addComponent(service: Payload) {
    const exist = this.components.find(
      (s) =>
        s.name === service.name ||
        (s.tech && service.tech && s.tech === service.tech)
    );
    if (exist) {
      // TODO: merge
      return;
    }

    this.components.push(service);
    for (const tech of service.techs) {
      this.techs.add(tech);
    }

    if (service.tech) {
      this.techs.add(service.tech);
    }
  }

  addTech(tech: AllowedKeys | AllowedKeys[]) {
    if (Array.isArray(tech)) {
      tech.forEach((t) => this.techs.add(t));
      return;
    }

    this.techs.add(tech);
  }

  addEdges(id: string) {
    this.edges.push({
      to: id,
      portSource: 'right',
      portTarget: 'left',
      read: true,
      write: true,
      vertices: [],
    });
  }

  detectLang(filename: string) {
    const ext = path.extname(filename);

    for (const lang of languages) {
      if (!lang.extensions.includes(ext)) {
        continue;
      }

      this.addLang(lang.group || lang.name);
      return;
    }

    for (const lang of others) {
      if (!lang.extensions.includes(ext)) {
        continue;
      }

      this.addLang(lang.group || lang.name);
      return;
    }
  }

  merge(pl: Payload) {
    this.addTech([...pl.techs]);
    pl.components.forEach((component) => this.addComponent(component));

    if (pl.name !== 'virtual') {
      this.addComponent(pl);
    }

    for (const [lang, count] of Object.entries(pl.languages)) {
      this.addLang(lang, count);
    }
  }

  toJson(): TechAnalyser {
    return {
      id: this.id,
      name: this.name,
      group: this.group,
      path: this.path,
      tech: this.tech,
      edges: this.edges,
      inComponent: this.inComponent,
      components: this.components.map((service) => {
        const { components, ...rest } = service.toJson();
        return rest;
      }),
      techs: [...this.techs].sort(),
      languages: this.languages,
    };
  }

  private addLang(name: string, count: number = 1) {
    if (!this.languages[name]) {
      this.languages[name] = 0;
    }

    this.languages[name] += count;

    if (name in nameToKey) {
      this.addTech(nameToKey[name]);
    }
  }
}
