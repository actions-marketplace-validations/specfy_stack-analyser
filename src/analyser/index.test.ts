import path from 'node:path';

import { describe, it, expect } from 'vitest';

import { Payload } from '../payload';
import { flatten } from '../payload/helpers';
import { FakeProvider } from '../provider/fake';
import { FSProvider } from '../provider/fs';
import { cleanSnapshot } from '../tests/helpers';
import type { TechAnalyser } from '../types';

import { techAnalyser } from '.';

const dockerCompose = `version: '3'
services:
  db:
    container_name: db
    image: postgres:15.1-alpine
    ports:
      - '5432:5432'
    environment:
      - POSTGRES_PASSWORD=postgres
`;

describe('techAnalyser', () => {
  it('should not find anything', async () => {
    const res = await techAnalyser({
      provider: new FakeProvider({
        paths: {
          '/': [],
        },
        files: {},
      }),
    });

    expect(res.toJson()).toStrictEqual({
      id: expect.any(String),
      name: 'main',
      group: 'component',
      edges: [],
      inComponent: null,
      languages: {},
      path: ['/'],
      tech: null,
      techs: [],
      childs: [],
    });
  });

  it('should register only component of the same tech', async () => {
    const res = await techAnalyser({
      provider: new FakeProvider({
        paths: {
          '/': ['package.json', 'docker-compose.yml'],
        },
        files: {
          '/docker-compose.yml': dockerCompose,
          '/package.json': '{ "name": "test", "dependencies": {"pg": "1.0.0"}}',
        },
      }),
    });

    const flat = new Payload({ name: 'flat', folderPath: '/' });
    flatten(res, flat);

    const json: TechAnalyser = JSON.parse(JSON.stringify(flat));
    expect(json).toStrictEqual({
      childs: [
        {
          childs: [],
          dependencies: {},
          edges: [],
          group: 'component',
          id: expect.any(String),
          inComponent: null,
          languages: {},
          name: 'db',
          parent: null,
          path: ['/docker-compose.yml', '/package.json'],
          tech: 'postgresql',
          techs: {},
        },
        {
          childs: [],
          dependencies: {
            pg: '1.0.0',
          },
          edges: [
            {
              portSource: 'right',
              portTarget: 'left',
              read: true,
              to: expect.any(String),
              vertices: [],
              write: true,
            },
          ],
          group: 'component',
          id: expect.any(String),
          inComponent: null,
          languages: {
            JSON: 1,
            YAML: 1,
          },
          name: 'test',
          parent: null,
          path: ['/package.json'],
          tech: null,
          techs: {},
        },
        {
          childs: [],
          dependencies: {},
          edges: [],
          group: 'component',
          id: expect.any(String),
          inComponent: null,
          languages: {},
          name: 'main',
          parent: null,
          path: ['/'],
          tech: null,
          techs: {},
        },
      ],
      dependencies: {},
      edges: [],
      group: 'component',
      id: expect.any(String),
      inComponent: null,
      languages: {
        JSON: 1,
        YAML: 1,
      },
      name: 'flat',
      path: ['/'],
      tech: null,
      techs: {},
    });

    expect(json.childs[0].id).toEqual(json.childs[1].edges[0].to);
  });

  it('should run correctly', async () => {
    const res = await techAnalyser({
      provider: new FSProvider({
        path: path.join(__dirname, '../../tests/__fixtures__'),
      }),
    });

    expect(res.toJson()).toStrictEqual({
      childs: [
        {
          id: expect.any(String),
          name: 'db',
          group: 'component',
          edges: [],
          languages: {},
          path: ['/docker-compose.yml'],
          tech: 'postgresql',
          techs: [],
          inComponent: null,
          childs: [],
        },
        {
          id: expect.any(String),
          name: 'fake',
          edges: [],
          group: 'component',
          path: ['/package.json'],
          inComponent: null,
          tech: null,
          languages: {
            HCL: 1,
            JSON: 1,
            YAML: 1,
          },
          techs: [
            'docker',
            'eslint',
            'nodejs',
            'prettier',
            'terraform',
            'typescript',
          ],
          childs: [
            {
              id: expect.any(String),
              name: '@fake/api',
              edges: [],
              group: 'component',
              path: ['/pkgs/api/package.json'],
              tech: null,
              inComponent: null,
              languages: {
                JSON: 1,
              },
              techs: ['fastify', 'nodejs', 'prisma', 'typescript'],
              childs: [],
            },
            {
              id: expect.any(String),
              name: '@fake/app',
              edges: [],
              group: 'component',
              path: ['/pkgs/app/package.json'],
              inComponent: expect.any(String),
              tech: null,
              languages: {
                HTML: 1,
                JSON: 1,
                SCSS: 1,
              },
              techs: [
                'html',
                'nodejs',
                'react',
                'scss',
                'typescript',
                'vercel',
                'vite',
              ],
              childs: [
                {
                  id: expect.any(String),
                  name: 'vercel',
                  group: 'hosting',
                  languages: {},
                  edges: [],
                  path: ['/pkgs/app/package.json'],
                  tech: 'vercel',
                  inComponent: null,
                  techs: [],
                  childs: [],
                },
              ],
            },
            {
              id: expect.any(String),
              name: 'GCP',
              group: 'hosting',
              languages: {},
              edges: [],
              path: ['/terraform/.terraform.lock.hcl'],
              tech: 'gcp',
              inComponent: null,
              techs: [],
              childs: [],
            },
            {
              id: expect.any(String),
              name: 'Vercel',
              group: 'hosting',
              languages: {},
              edges: [],
              path: ['/terraform/.terraform.lock.hcl'],
              tech: 'vercel',
              inComponent: null,
              techs: [],
              childs: [],
            },
          ],
        },
        {
          id: expect.any(String),
          name: 'redis',
          group: 'component',
          edges: [],
          languages: {},
          path: ['/docker-compose.yml'],
          tech: 'redis',
          techs: [],
          inComponent: null,
          childs: [],
        },
        {
          id: expect.any(String),
          name: 'unknown',
          group: 'component',
          edges: [],
          languages: {},
          path: ['/docker-compose.yml'],
          tech: null,
          techs: [],
          inComponent: null,
          childs: [],
        },
      ],
      edges: [],
      group: 'component',
      id: expect.any(String),
      inComponent: null,
      languages: {},
      name: 'main',
      path: ['/'],
      tech: null,
      techs: [],
    });

    const flatted = flatten(res);
    expect(
      cleanSnapshot(JSON.parse(JSON.stringify(flatted.toJson())))
    ).toMatchSnapshot();
  });
});
