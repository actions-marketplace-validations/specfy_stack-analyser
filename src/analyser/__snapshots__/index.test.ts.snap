// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`techAnalyser > should run correctly 1`] = `
{
  "childs": [
    {
      "childs": [],
      "edges": [],
      "group": "component",
      "id": "",
      "inComponent": null,
      "languages": {
        "JSON": 1,
      },
      "name": "@fake/api",
      "path": [
        "/pkgs/api/package.json",
      ],
      "tech": null,
      "techs": [
        "fastify",
        "nodejs",
        "prisma",
        "typescript",
      ],
    },
    {
      "childs": [],
      "edges": [],
      "group": "component",
      "id": "",
      "inComponent": "",
      "languages": {
        "HTML": 1,
        "JSON": 1,
        "SCSS": 1,
      },
      "name": "@fake/app",
      "path": [
        "/pkgs/app/package.json",
      ],
      "tech": null,
      "techs": [
        "html",
        "nodejs",
        "react",
        "scss",
        "typescript",
        "vercel",
        "vite",
      ],
    },
    {
      "childs": [],
      "edges": [],
      "group": "hosting",
      "id": "",
      "inComponent": null,
      "languages": {},
      "name": "GCP",
      "path": [
        "/terraform/.terraform.lock.hcl",
      ],
      "tech": "gcp",
      "techs": [],
    },
    {
      "childs": [],
      "edges": [],
      "group": "component",
      "id": "",
      "inComponent": null,
      "languages": {},
      "name": "db",
      "path": [
        "/docker-compose.yml",
      ],
      "tech": "postgresql",
      "techs": [],
    },
    {
      "childs": [],
      "edges": [],
      "group": "component",
      "id": "",
      "inComponent": null,
      "languages": {
        "HCL": 1,
        "JSON": 1,
        "YAML": 1,
      },
      "name": "fake",
      "path": [
        "/package.json",
      ],
      "tech": null,
      "techs": [
        "docker",
        "eslint",
        "nodejs",
        "prettier",
        "terraform",
        "typescript",
      ],
    },
    {
      "childs": [],
      "edges": [],
      "group": "component",
      "id": "",
      "inComponent": null,
      "languages": {},
      "name": "main",
      "path": [
        "/",
      ],
      "tech": null,
      "techs": [],
    },
    {
      "childs": [],
      "edges": [],
      "group": "component",
      "id": "",
      "inComponent": null,
      "languages": {},
      "name": "redis",
      "path": [
        "/docker-compose.yml",
      ],
      "tech": "redis",
      "techs": [],
    },
    {
      "childs": [],
      "edges": [],
      "group": "component",
      "id": "",
      "inComponent": null,
      "languages": {},
      "name": "unknown",
      "path": [
        "/docker-compose.yml",
      ],
      "tech": null,
      "techs": [],
    },
    {
      "childs": [],
      "edges": [],
      "group": "hosting",
      "id": "",
      "inComponent": null,
      "languages": {},
      "name": "vercel",
      "path": [
        "/pkgs/app/package.json",
        "/terraform/.terraform.lock.hcl",
      ],
      "tech": "vercel",
      "techs": [],
    },
  ],
  "edges": [],
  "group": "component",
  "id": "",
  "inComponent": null,
  "languages": {
    "HCL": 1,
    "HTML": 1,
    "JSON": 3,
    "SCSS": 1,
    "YAML": 1,
  },
  "name": "flatten",
  "path": [
    "/",
  ],
  "tech": null,
  "techs": [
    "docker",
    "eslint",
    "fastify",
    "gcp",
    "html",
    "nodejs",
    "postgresql",
    "prettier",
    "prisma",
    "react",
    "redis",
    "scss",
    "terraform",
    "typescript",
    "vercel",
    "vite",
  ],
}
`;
